### PR TITLE
Avoid ifdefs in trace categories

### LINKED ISSUE #19915: Trace: Index error in trace_attach_cb and other functions
With openssl 3.0.7 (other versions not tested) there is an out-of-bound access to array ```trace_categories``` in ```trace_attach_cb```, ```trace_detach_cb``` and ```trace_attach_w_callback_cb```:

```
        OSSL_TRACE2(TRACE, "Attach channel %p to category '%s'\n",
                    data, trace_categories[category].name);
```

The trace_categories array is defined as:

```
static const struct trace_category_st trace_categories[] = {
    TRACE_CATEGORY_(ALL),
    ...
TRACE_CATEGORY_(CONF),
#ifndef OPENSSL_NO_ENGINE
    TRACE_CATEGORY_(ENGINE_TABLE),
    TRACE_CATEGORY_(ENGINE_REF_COUNT),
#endif
    TRACE_CATEGORY_(PKCS5V2),
    ...
    TRACE_CATEGORY_(REF_COUNT)
};
```

Its length depends on whether ```OPENSSL_NO_ENGINE``` is defined or not. If it is not defined, the categories numbers are running from 0 to 17 with ```OSSL_TRACE_CATEGORY_NUM``` defined as 18 in trace.h, while the valid array indices are running from 0 to 15. Loops are correctly using ```OSSL_NELEM(trace_categories)``` as upper bound (resulting in a value of 16).

Accordingly, you must never access the array ```trace_categories``` with the category-number itself like in
 
```
        OSSL_TRACE2(TRACE, "Attach channel %p to category '%s'\n",
                    data, trace_categories[category].name);
```

As a solution, I've added an index-mapping function ```OSSL_trace_get_category_index``` in trace.c:

```
static int OSSL_trace_get_category_index(int num)
{
    size_t i;

    for (i = 0; i < OSSL_NELEM(trace_categories); i++)
        if (trace_categories[i].num == num)
            return (int)i;
    return -1; /* not found */
}

...

static int trace_attach_cb(int category, int type, const void *data)
{
int category_idx = OSSL_trace_get_category_index( category );

    if ( category_idx < 0 )
        return 0;

    switch (type) {
    case CHANNEL:
        OSSL_TRACE2(TRACE, "Attach channel %p to category '%s'\n",
                    data, trace_categories[category_idx].name);
    ...
```

The new function retrieves the array index of the entry with a given category.

### PR DESCRIPTION
**Pull request by t8m**
The trace code assumes all categories are present and
the category numbers are equal to the index in the table.
    
Fixes #19915

##### Checklist

- [x] tests are added or updated

### CODE REVIEW

File: test/trace_api_test.c [hunk 1]
```diff
@@ -69,10 +67,10 @@ static int test_trace_categories(void)
 #ifndef OPENSSL_NO_TRACE
 static void put_trace_output(void)
 {
-    OSSL_TRACE_BEGIN(TLS) {
+    OSSL_TRACE_BEGIN(HTTP) {
         BIO_printf(trc_out, "Hello World\n");
         BIO_printf(trc_out, "Good Bye Universe\n");
-    } OSSL_TRACE_END(TLS);
+    } OSSL_TRACE_END(HTTP);
```
Comment (levitte) [inline test/trace_api_test.c:73]: ?

File: test/trace_api_test.c [hunk 2]
```diff
@@ -88,30 +86,60 @@ static int test_trace_channel(void)
...
-    if (!TEST_int_eq(OSSL_trace_set_suffix(OSSL_TRACE_CATEGORY_TLS, "-abc"), 1))
+    if (!TEST_int_eq(OSSL_trace_set_suffix(OSSL_TRACE_CATEGORY_HTTP, "-abc"), 1))
         goto end;
 
     put_trace_output();
     len = BIO_get_mem_data(bio, &p_buf);
     if (!TEST_strn2_eq(p_buf, len, expected, expected_len))
         goto end;
-    if (!TEST_int_eq(OSSL_trace_set_channel(OSSL_TRACE_CATEGORY_TLS, NULL), 1))
+    if (!TEST_int_eq(OSSL_trace_set_channel(OSSL_TRACE_CATEGORY_HTTP, NULL), 1))
```
Comment (levitte) [inline test/trace_api_test.c:104]: ?

**Review by levitte — COMMENTED** [review]
Any reason for the change from TLS to HTTP trace category?

**Comment by t8m** [comment]
> Any reason for the change from TLS to HTTP trace category?

The HTTP category is after the ifdefed out TLS category (in fact it is the last one). So this is needed to trigger the ASAN failure in the test without the fix.

**Comment by levitte** [comment]
> > Any reason for the change from TLS to HTTP trace category?
> 
> The HTTP category is after the ifdefed out TLS category (in fact it is the last one). So this is needed to trigger the ASAN failure in the test without the fix.

Ah ok, makes sense

File: include/openssl/trace.h [hunk 3]
```diff
@@ -43,10 +43,8 @@ extern "C" {
 # define OSSL_TRACE_CATEGORY_TLS                 3
 # define OSSL_TRACE_CATEGORY_TLS_CIPHER          4
 # define OSSL_TRACE_CATEGORY_CONF                5
-# ifndef OPENSSL_NO_ENGINE
-#  define OSSL_TRACE_CATEGORY_ENGINE_TABLE       6
-#  define OSSL_TRACE_CATEGORY_ENGINE_REF_COUNT   7
-# endif
+# define OSSL_TRACE_CATEGORY_ENGINE_TABLE        6
```
Comment (slontis) [inline include/openssl/trace.h:46]: Shifting the indexes to the end would of probably also worked :)

Comment (t8m) [reply include/openssl/trace.h:46]: You mean the engine ones? But we cannot do that as this is a public define.

**Review by slontis — APPROVED** [review]

**Review by beldmit — APPROVED** [review]
LGTM

**Comment by t8m** [comment]
Merged to master and 3.1 branches. 3.0 will require a separate PR.

Thank you for the reviews.

**Pull request closed without merging**